src/corefx/System.Globalization.Native/collation.cpp

```diff
@@ -0,0 +1,254 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+#include <assert.h>
+#include <stdint.h>
+#include <unicode/uchar.h>
+#include <unicode/ucol.h>
+#include <unicode/usearch.h>
+#include <unicode/utf16.h>
+
+const int32_t CompareOptionsIgnoreCase = 1;
+// const int32_t CompareOptionsIgnoreNonSpace = 2;
+// const int32_t CompareOptionsIgnoreSymbols = 4;
+// const int32_t CompareOptionsIgnoreKanaType = 8;
+// const int32_t CompareOptionsIgnoreWidth = 0x10;
+// const int32_t CompareOptionsStringSort = 0x20000000;
+
+/*
+ * To collator returned by this function is owned by the callee and must be closed when this method returns
+ * with a U_SUCCESS UErrorCode.
+ *
+ * On error, the return value is undefined.
+ */
+UCollator* GetCollatorForLocaleAndOptions(const char* lpLocaleName, int32_t options, UErrorCode* pErr)
+{
+    UCollator* pColl = nullptr;
+
+    pColl = ucol_open(lpLocaleName, pErr);
+
+    if ((options & CompareOptionsIgnoreCase) == CompareOptionsIgnoreCase)
+    {
+        ucol_setAttribute(pColl, UCOL_STRENGTH, UCOL_SECONDARY, pErr);
+    }
+
+    return pColl;
+}
+
+
+/*
+Function:
+CompareString
+*/
+extern "C" int32_t CompareString(const char* lpLocaleName, const UChar* lpStr1, int32_t cwStr1Length, const UChar* lpStr2, int32_t cwStr2Length, int32_t options)
+{
+    UCollationResult result = UCOL_EQUAL;
+    UErrorCode err = U_ZERO_ERROR;
+    UCollator* pColl = GetCollatorForLocaleAndOptions(lpLocaleName, options, &err);
+
+    if (U_SUCCESS(err))
+    {
+        result = ucol_strcoll(pColl, lpStr1, cwStr1Length, lpStr2, cwStr2Length);
+        ucol_close(pColl);
+    }
+
+    return result;
+}
+
+/*
+Function:
+IndexOf
+*/
+extern "C" int32_t IndexOf(const char* lpLocaleName, const UChar* lpTarget, const UChar* lpSource, int32_t cwSourceLength, int32_t options)
+{
+    int32_t result = USEARCH_DONE;
```

Comment:
It looks like the managed caller is assuming USEARCH_DONE is -1.  Should we static_assert that in the shim so that we know at compile time if those assumptions are violated?  Same goes for some of the other result values.